### PR TITLE
[wasm-ep] Implement DiagnosticServer and startup sessions for WebAssembly

### LINKED ISSUE #69674: [wasm-ep] Add a hook to start an EventPipe Session from runtime startup
Required for wasm profiled AOT

### PR DESCRIPTION
**Pull request by lambdageek**
Add a diagnostic server for WebAssembly.  Enable by building the runtime with `/p:WasmEnablePerfTracing=true` or `/p:WasmEnableThreads=true`.

To configure a project to start the diagnostic server, add this to the .csproj:
```xml
      <WasmExtraConfig Include="diagnostic_options" Value='
{
  "server": { "suspend": false, "connect_url": "ws://localhost:8088/diagnostics" }
}' />
```

The `connect_url` should be a WebSocket url serviced by `dotnet-dsrouter server-websocket` **from this branch** https://github.com/lambdageek/diagnostics/tree/wasm-server

Note that setting `"suspend": true` will hang the browser tab until a diagnostic tool such as `dotnet-trace collect` connects to the dsrouter.

---

Implement creating VFS file based sessions at runtime startup.  Add the following to a .csproj:

```xml
    <WasmExtraConfig Include="diagnostic_options" Value='
{
  "sessions": [ { "collectRundownEvents": "true", "providers": "WasmHello::5:EventCounterIntervalSec=1" } ]
}' />
```

That will create and start one or more EventPipe sessions that will store their results into the VFS.

The startup session can be retrieved via `MONO.diagnostics.getStartupSessions()`.  Each session `s` should be stopped via `s.stop()` and the data can then be extraced in a `Blob` using `s.getTraceBlob()`.

This is orthogonal to the diagnostic server support.  You don't need `dotnet-dsrouter` running on the host.  But you do need access to JavaScript on the main thread.

---

Notes/Issues:

* Tree shaking: I verified that if threads are not available, all the TypeScript diagnostics code is removed.
* Right now the server is not very robust to `dotnet-dsrouter` stopping, or starting after the runtime starts.  The ideal order is to start `dotnet-dsrouter` first, and then open the browser
* Unrelated housekeeping fixes:
   * Tell `wasm.proj` about all the subdirectories with .ts files - makes incremental builds notice changes in subdirectories.
   * Add a rollup `dependencies` property to quiet a warning about `node/buffer`
   * There's a mock implementation of a "websocket" that was used for protocol testing. I verified that tree-shaking removes this in thread-enabled Release builds.
   * Bump `PTHREAD_POOL_SIZE` to `4` and set `PTHREAD_POOL_SIZE_STRICT=2` (returns `EAGAIN` from `pthread_create` if the pool needs to grow).  The previous setting `PTHREAD_POOL_SIZE_STRING=1` (warn and try to grow the pool) seemed to lead to hangs.  Usually that means the main thread is creating a thread and immediately calling `pthread_join` without returning to JS. We should investigate separately.
   * The only implemented diagnostic server commands are `CollectTracing2`, `StopCollecting` and `ResumeRuntime`. None of the `Dump`, `Process` and `Profiler` commands are implemented and the server will crash if it receives them.  It should be relatively straightforward to return a "command unsupported" reply (which would allow clients to gracefully disconnect), but it's not done yet.
* In some error states the runtime kills the browser tab with the following in a terminal window (if Chrome is started from a terminal: `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory`). This probably means we're hitting a loop somewhere that rapidly exhausts JIT memory, but it's difficult to investigate when the JS console dies, too (happens with chrome stable v103 and chrome beta v104).

Fixes https://github.com/dotnet/runtime/issues/69674, contributes to https://github.com/dotnet/runtime/issues/72481

### CODE REVIEW

**Comment by lambdageek** [comment]
@pavelsavara this is going to have a merge conflict with #72275 in `startup.ts` (and probably other minor ones).
The main thing that I changed was to make `_apply_configuration_from_args` into an async function (so that I can call the async `mono_wasm_diagnostics_init` which needs to make the main thread wait until the diagnostic server thread starts.
There is probably some way to do the same thing in the new setup?  Is it possible to make `mono_wasm_before_user_runtime_initialized` async?

File: src/mono/wasm/runtime/workers/README.md
Comment (pavelsavara) [inline src/mono/wasm/runtime/workers/README.md:24]: Could you please describe how do you enable the mocks and how to use them for testing ? Is there unit test which uses it ?

**Comment by pavelsavara** [comment]
> I tested merging the branches. Looks like the only thing I had to do was to make `mono_wasm_before_user_runtime_initialized` async.

Good, I don't know if I would merge my PR in Net7, probably just smaller portion of it. So feel free to merge as is.
Looked at the typescript and LGTM.
It would be good to have JavaScript-only unit tests for complex components at some point. https://jestjs.io/ is good choice I think.

File: src/mono/mono/component/event_pipe.c
Comment (lateralusX) [inline src/mono/mono/component/event_pipe.c:366]: Remove IpcStream comment below, since it has been moved.

**Review by lateralusX — COMMENTED** [review]
Native EventPipe bits LGTM!

File: src/mono/mono/component/diagnostics_server.c
Comment (lambdageek) [reply src/mono/mono/component/diagnostics_server.c:97]: I used to print a warning to the JS console here, but it's not actually that useful yet because it's difficult to access the JS console while the browser tab is busy.

I'll simplify this for now.

File: src/mono/mono.proj
Comment (lambdageek) [reply src/mono/mono.proj:486]: yea the idea was to disable most of the C diagnostic server functionality as it is not needed.

File: src/mono/wasm/runtime/startup.ts
Comment (pavelsavara) [inline src/mono/wasm/runtime/startup.ts:797]: `MONO_WASM: ` prefix

File: src/mono/wasm/runtime/workers/README.md
Comment (lambdageek) [reply src/mono/wasm/runtime/workers/README.md:24]: Just to update the status: I'm working on a couple of changes to the mock stuff:
1. Move the actual mock protocol script into the sample project.  The plan is is that the .csproj `diagnostic_options.server_url` json can be `mock:relative_url/mock.js` and then we will `import(mock_url).then(module => mock(module.default).run()` - ie: the mock script is per-project and it's just an es6 module with a default export in a format that the mock engine understands.
2. I'm adding some utilities to the mock engine so it can say things like `connection.expectOkReply ((payload) => /*do a check of the payload */)` rather than leaving the mock script to parse the bytes on its own.
3. document how to use it.

Should be done by EOD today.

**Comment by lambdageek** [comment]
> It would be good to have JavaScript-only unit tests for complex components at some point. https://jestjs.io/ is good choice I think.

I'll play around with that this afternoon.  Probably won't be part of this PR unless it turns out to be completely trivial to add the unit tests to our setup.

File: src/mono/wasm/runtime/workers/README.md
Comment (lambdageek) [reply src/mono/wasm/runtime/workers/README.md:24]: ok, mocking changes are up

Comment (lambdageek) [reply src/mono/wasm/runtime/workers/README.md:24]: I also generate `diagnostics-mock.d.ts` into the `wasm/runtime` directory in debug builds, same as `dotnet.d.ts`. I found it helpful to have TypeScript check my JS mocking script.

**Comment by lambdageek** [comment]
verified one more time that with threading or perftracing disabled, all of the related JS is removed by tree-shaking

File: src/mono/wasm/runtime/diagnostics/README.md
Comment (pavelsavara) [inline src/mono/wasm/runtime/diagnostics/README.md:13]: What is it we are mocking and why ? (high level intro)

Comment (pavelsavara) [reply src/mono/wasm/runtime/diagnostics/README.md:13]: Do I understand that the diagnosed process is a "server" and the tooling outside of the browser is a "client" ? And so, we are mocking the client here and it's requests ?

File: src/mono/wasm/runtime/diagnostics/shared/create-session.ts
Comment (pavelsavara) [inline src/mono/wasm/runtime/diagnostics/shared/create-session.ts:42]: getU32 for pointers

File: src/mono/wasm/runtime/polyfills.ts
Comment (pavelsavara) [inline src/mono/wasm/runtime/polyfills.ts:102]: do we need copy of `subscribers` because of the change in the loop ?

File: src/mono/wasm/runtime/pthreads/worker/index.ts
Comment (pavelsavara) [inline src/mono/wasm/runtime/pthreads/worker/index.ts:57]: After we merge both this and my PR, we could prefix this with `if (runtimeHelpers.diagnostic_tracing)` so that by default the runtime is not noisy. It matters in console like nodeJS. @maraf suggested it's about a time when we introduce logging component to centralize all this logic to one place.

**Review by pavelsavara — APPROVED** [review]
LGTM, none of my other comments are blocking merge.

File: src/mono/wasm/runtime/diagnostics/server_pthread/mock-remote.ts
Comment (lambdageek) [reply src/mono/wasm/runtime/diagnostics/server_pthread/mock-remote.ts:13]: Yea, if there's a need for writing the mocking script using CJS, we can add it here.  But since the goal here is just to have a tool to improve the server's robustness (for example: make scripts that send garbage and verify that the DS can recover), I don't think there's a need for CJS.  This isn't something we're ever going to ship. It's just for better testing.

File: src/mono/wasm/runtime/pthreads/worker/index.ts
Comment (lambdageek) [reply src/mono/wasm/runtime/pthreads/worker/index.ts:57]: agreed

**Comment by lambdageek** [comment]
https://helix.dot.net/api/2019-06-17/jobs/d29bfb53-4790-47f4-8a1e-aa868ec7afd5/workitems/System.Net.Mail.Functional.Tests/console failure is unrelated

```
/__w/1/s/src/native/libs/Common/pal_utilities.h:86: int ToFileDescriptor(intptr_t): Assertion `0 <= fd && fd < sysconf(_SC_OPEN_MAX)' failed.
```

**Pull request merged**
